File: deduced-self/printer.py
from abc import ABC, abstractmethod
from typing import TypeVar


TDerived = TypeVar('TDerived', bound='Printer')


class Printer(ABC):
    def print_line(self: TDerived, message: str) -> TDerived:
        self._write(message + '\n')
        return self

    @abstractmethod
    def _write(self, message: str) -> None:
        raise NotImplementedError()


class StdoutPrinter(Printer):
    def print_fancy(self, message: str) -> 'StdoutPrinter':
        return self.print_line(f'~~~ {message} ~~~')

    def _write(self, message: str) -> None:
        print(message, end='')


printer = StdoutPrinter()
printer.print_line('hello world').print_fancy('Deduced self works!')

File: deduced-self/printer.rs
trait Printer {
    fn print_line(&self, message: &str) -> &Self {
        self._write(&format!("{message}\n"));
        return self;
    }

    fn _write(&self, message: &str);
}


struct StdoutPrinter {}

impl StdoutPrinter {
    fn print_fancy(&self, message: &str) -> &Self {
        return self.print_line(&format!("~~~ {message} ~~~"));
    }
}

impl Printer for StdoutPrinter {
    fn _write(&self, message: &str) {
        print!("{message}");
    }
}


fn main() {
    let printer = StdoutPrinter{};
    printer.print_line("hello world").print_fancy("Traits work!");
}
